File: buildSrc/src/main/java/org/briarproject/briar/desktop/builddata/BuildDataPlugin.java
package org.briarproject.briar.desktop.builddata;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.logging.Logger;
import org.gradle.api.plugins.JavaPluginConvention;
import org.gradle.api.tasks.SourceSet;

import java.nio.file.Path;

public class BuildDataPlugin implements Plugin<Project> {

	@Override
	public void apply(final Project project) {
		Logger logger = project.getLogger();
		logger.info("applying version access plugin");

		BuildDataPluginExtension extension = project.getExtensions().create(
				"buildData", BuildDataPluginExtension.class);

		GenerateBuildDataSourceTask task = project.getTasks().create(
				"buildData", GenerateBuildDataSourceTask.class);
		task.setConfiguration(extension);

		project.getTasks().findByName("compileJava").dependsOn(task);

		Path pathBuildDir = project.getBuildDir().toPath();
		Path source = Util.getSourceDir(pathBuildDir);

		SourceSet sourceSets = project.getConvention()
				.getPlugin(JavaPluginConvention.class).getSourceSets()
				.findByName("main");
		sourceSets.java(sourceSet -> {
			sourceSet.srcDir(source);
		});
	}

}

File: buildSrc/src/main/kotlin/org/briarproject/briar/desktop/builddata/BuildDataPlugin.kt
/*
 * Briar Desktop
 * Copyright (C) 2021-2022 The Briar Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.briarproject.briar.desktop.builddata

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.file.SourceDirectorySet
import org.gradle.api.plugins.JavaPluginExtension

class BuildDataPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        val logger = project.logger
        logger.info("applying version access plugin")

        val extension = project.extensions.create(
            "buildData", BuildDataPluginExtension::class.java
        )
        val task = project.tasks.create(
            "buildData", GenerateBuildDataSourceTask::class.java
        )
        task.configuration = extension
        project.tasks.findByName("compileJava")!!.dependsOn(task)
        val pathBuildDir = project.buildDir.toPath()
        val source = Util.getSourceDir(pathBuildDir)
        val sourceSets = project.extensions.getByType(JavaPluginExtension::class.java).sourceSets.findByName("main")
        sourceSets!!.java { sourceSet: SourceDirectorySet -> sourceSet.srcDir(source) }
    }
}
